behaviour/graql/reasoner/relation-inference.feature

```diff
@@ -0,0 +1,1299 @@
+      nutella-price sub rule,
+      when {
+        $x isa item, has name "3kg jar of Nutella";
+        $y 14.99 isa price;
+      }, then {
+        (listed-item: $x, item-price: $y) isa item-listing;
+      };
+      """
+    Given for each session, graql insert
+      """
+      insert
+      $x isa item, has name "3kg jar of Nutella";
+      $y 14.99 isa price;
+      """
+    When materialised keyspace is completed
+    Then for graql query
+      """
+      match
+        $r (listed-item: $i, item-price: $p) isa item-listing;
+        $i isa item, has name $n;
+        $n "3kg jar of Nutella" isa name;
+        $p 14.99 isa price;
+      get;
+      """
+    Then all answers are correct in reasoned keyspace
+    Then answer size in reasoned keyspace is: 1
+    Then materialised and reasoned keyspaces are the same size
+
+
+  Scenario: a rule can infer a relation based on ownership of any instance of a specific attribute type
+    Given for each session, graql define
+      """
+      define
+      year sub attribute, value long, plays favourite-year;
+      employment relates favourite-year;
+      kronenbourg-employs-anyone-with-a-name sub rule,
+      when {
+        $x isa company, has name "Kronenbourg";
+        $p isa person, has name $n;
+        $y 1664 isa year;
+      }, then {
+        (employee: $p, employer: $x, favourite-year: $y) isa employment;
+      };
+      """
+    Given for each session, graql insert
+      """
+      insert
+      $x isa company, has name "Kronenbourg";
+      $p isa person, has name "Ronald";
+      $p2 isa person, has name "Prasanth";
+      $y 1664 isa year;
+      """
+    When materialised keyspace is completed
+    Then for graql query
+      """
+      match
+        $x 1664 isa year;
+        ($x, employee: $p, employer: $y) isa employment;
+      get;
+      """
+    Then all answers are correct in reasoned keyspace
+    Then answer size in reasoned keyspace is: 2
+    Then materialised and reasoned keyspaces are the same size
+
+
+  ###############
+  # REFLEXIVITY #
+  ###############
+
+  Scenario: when inferring relations on all pairs from n concepts, the number of relations is the nth triangle number
+    Given for each session, graql define
+      """
+      define
+      everyone-is-my-friend-including-myself sub rule,
+      when {
+        $x isa person;
+        $y isa person;
+      }, then {
+        (friend: $x, friend: $y) isa friendship;
+      };
+      """
+    Given for each session, graql insert
+      """
+      insert
+      $a isa person, has name "Abigail";
+      $b isa person, has name "Bernadette";
+      $c isa person, has name "Cliff";
+      $d isa person, has name "Damien";
+      $e isa person, has name "Eustace";
+      """
+    When materialised keyspace is completed
+    Then for graql query
+      """
+      match $r isa friendship; get;
+      """
+    Then all answers are correct in reasoned keyspace
+    Then answer size in reasoned keyspace is: 15
```

Comment:
maybe "number of relations is the number of unordered pairs of players = n *(n-1)/2"

Reply:
I went a bit further.
```
# When there is 1 concept we have {aa}.
    # Adding a 2nd concept gives us 2 new relations - where each relation contains b, and one other concept (a or b).
    # Adding a 3rd concept gives us 3 new relations - where each relation contains c, and one other concept (a, b or c).
    # Generally, the total number of relations is the sum of all integers from 1 to n inclusive.
```